src/allocator_defrag.c

```diff
@@ -0,0 +1,476 @@
+
+
+/* -----------------------------------------------------------------------------
+ * Alloc/Free API that are cooperative with defrag
+ * -------------------------------------------------------------------------- */
+/* Allocation and free functions that bypass the thread cache
+ * and go straight to the allocator arena bins.
+ * Currently implemented only for jemalloc. Used for online defragmentation. */
+void *allocatorDefragAlloc(size_t size) {
+    void *ptr = je_mallocx(size, MALLOCX_TCACHE_NONE);
+    return ptr;
+}
+void allocatorDefragFree(void *ptr, size_t size) {
+    if (ptr == NULL) return;
+    je_sdallocx(ptr, size, MALLOCX_TCACHE_NONE);
+}
+
+/* -----------------------------------------------------------------------------
+ * Helper functions for jemalloc translation between size and index
+ * -------------------------------------------------------------------------- */
+#define LG_QUANTOM_8_FIRST_POW2 3
+#define SIZE_CLASS_GROUP_SZ 4
+
+#define LG_QUANTOM_OFFSET_3 ((64 >> LG_QUANTOM_8_FIRST_POW2) - 1)
+#define LG_QUANTOM_OFFSET_4 (64 >> 4)
+
+#define getBinindNormal(_sz, _offset, _last_sz_pow2)                                                             \
+    ((SIZE_CLASS_GROUP_SZ - (((1 << (_last_sz_pow2)) - (_sz)) >> ((_last_sz_pow2) - LG_QUANTOM_8_FIRST_POW2))) + \
+     (((_last_sz_pow2) - (LG_QUANTOM_8_FIRST_POW2 + 3)) - 1) * SIZE_CLASS_GROUP_SZ + (_offset))
+/* Get the bin index in bin array from the reg_size.
+ *
+ * these are reverse engineered mapping of reg_size -> binind. We need this information because the utilization query
+ * returns the size of the buffer and not the bin index, and we need the bin index to access it's usage information
+ *
+ * Note: In case future PR will return the binind (that is better API anyway) we can get rid of
+ * these conversion functions*/
+inline unsigned jeSize2BinIndexLgQ3(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 24, 32, 40, 48, 56, 64
+        return (sz >> 3) - 1;
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return getBinindNormal(sz, LG_QUANTOM_OFFSET_3, last_sz_in_group_pow2);
+}
+
+/* -----------------------------------------------------------------------------
+ * Get INFO string about the defrag.
+ * -------------------------------------------------------------------------- */
+/*
+ * add defrag info string into info
+ */
+sds allocatorDefragCatFragmentationInfo(sds info) {
+    if (!defrag_supported) return info;
+    jeBinInfo *bin_info;
+    jeBusage *bin_usage;
+    unsigned nbins = arena_bin_conf.nbins;
+    if (nbins > 0) {
+        info = sdscatprintf(info,
+                            "jemalloc_quantum:%d\r\n"
+                            "defrag_hit_ratio:%.2f\r\n"
+                            "defrag_hits:%lu\r\n"
+                            "defrag_misses:%lu\r\n"
```

Comment:
How are these different from the existing stat_active_defrag_hits and misses.

Reply:
addressed this in previous [PR](https://github.com/valkey-io/valkey/pull/692#discussion_r1828912275), thoughts?

> I didn't want to change exiting defrag metrics tracked from defrag.c file. These (existing metrics) could be, potentially, for non jemalloc defrag and have upper level logic, like keys defragged. I can think of nice other metrics at the defrag.c level, like have information about types of keys that where defragged. Here this information is from the perspective of the allocator, this specific metric does correspond to the active_defrag_hits in existing code. @madolson Does this make sense? or do you still think we should consolidate?

Reply:
I didn't review the INFO fields until now.

I don't think we should add duplicate info or even a new section for this. If we ever support active defrag for another allocator, it can use the same info fields as we already have from before. We can add more info fields in the same section as the existing ones.

Under `# Memory` we have:

                "allocator_allocated:%zu\r\n", server.cron_malloc_stats.allocator_allocated,
                "allocator_active:%zu\r\n", server.cron_malloc_stats.allocator_active,
                "allocator_resident:%zu\r\n", server.cron_malloc_stats.allocator_resident,
                "allocator_muzzy:%zu\r\n", server.cron_malloc_stats.allocator_muzzy,

                "allocator_frag_ratio:%.2f\r\n", mh->allocator_frag,
                "allocator_frag_bytes:%zu\r\n", mh->allocator_frag_bytes,

                "mem_allocator:%s\r\n", ZMALLOC_LIB,

                "active_defrag_running:%d\r\n", server.active_defrag_running,

Under `# Stats` we have:

                "active_defrag_hits:%lld\r\n", server.stat_active_defrag_hits,
                "active_defrag_misses:%lld\r\n", server.stat_active_defrag_misses,
                "active_defrag_key_hits:%lld\r\n", server.stat_active_defrag_key_hits,
                "active_defrag_key_misses:%lld\r\n", server.stat_active_defrag_key_misses,
                "total_active_defrag_time:%lld\r\n", (server.stat_total_active_defrag_time + current_active_defrag_time) / 1000,
                "current_active_defrag_time:%lld\r\n", current_active_defrag_time / 1000,

IMO, the new fields you're adding here should go either under `# Stats` (for stats like hits and misses) or under `# Memory` for everything else. We already have things like `mem_allocator` and `active_defrag_running` under Memory.

Regarding the field `jemalloc_quantum:8`, can we rename it to something more generic, like `allocator_pointer_alignment:8` and display it for other allocators too?

Reply:
Ack, I think that at this point it's better to remove this stats all together, they seem to generate more confusion than value. We can re-add needed information when we see the need. I will remove entirely all the info section and the associated structs&code for updating this will simplify the PR overall. I am not sure anyone but me will use it at this stage, TBH :) 
@madolson @zuiderkwast , Does this make sense?

Reply:
please see update version